lib/galaxy/config/__init__.py

```diff
@@ -162,6 +172,78 @@ def _set_config_directories():
         _set_global_conf()
         _set_config_directories()
 
+    def _load_schema(self):
+        # Override in subclasses
+        raise Exception('Not implemented')
+
+    def _deprecated_dirs(self):
+        # Override in subclasses (optional)
+        # {KEY: config option, VALUE: deprecated directory name}
+        # If VALUE == first directory in a user path that resolves to KEY, it will be stripped from the path
+        return None
+
+    def _update_raw_config_from_kwargs(self, kwargs):
+
+        def convert_datatype(key, value):
+            datatype = self.schema.app_schema[key].get('type')
+            # check for `not None` explicitly (value can be falsy)
+            if value is not None and datatype in type_converters:
+                return type_converters[datatype](value)
+            return value
+
+        def strip_deprecated_dir(key, value):
+            resolves_to = self.schema.paths_to_resolve.get(key)
+            if resolves_to:  # value is a path that will be resolved
+                first_dir = value.split(os.sep)[0]  # get first directory component
+                if first_dir == self._deprecated_dirs().get(resolves_to):  # first_dir is deprecated for this option
+                    ignore = first_dir + os.sep
+                    log.warning(
+                        "Paths for the '%s' option are now relative to '%s', remove the leading '%s' "
+                        "to suppress this warning: %s", key, resolves_to, ignore, value
+                    )
+                    return value[len(ignore):]
+            return value
+
+        type_converters = {'bool': string_as_bool, 'int': int, 'float': float, 'str': str}
+
+        for key, value in kwargs.items():
+            if key in self.schema.app_schema:
+                value = convert_datatype(key, value)
+                if value and self._deprecated_dirs():
+                    value = strip_deprecated_dir(key, value)
+                self._raw_config[key] = value
+
+    def _create_attributes_from_raw_config(self):
+        for key, value in self._raw_config.items():
+            if not hasattr(self, key):
+                setattr(self, key, value)
+            else:
+                log.debug("Attribute '%s' is set and cannot be overwritten with value '%s'" % (key, value))
```

Comment:
Is it safe to just `log.debug()` in this case? It was raising a `ConfigurationError` exception before.

Reply:
In my opinion, it is not only safe but also necessary. Here're my thoughts: (commit message): https://github.com/galaxyproject/galaxy/pull/9371/commits/777379729198a893519ad58f167be761a4ea867d 
And here's our original discussion (you and I made it an exception in the first place): https://github.com/galaxyproject/galaxy/pull/8728#pullrequestreview-295263870

Reply:
The following are legit cases when this would happen: `config_dir`, `data_dir`, `managed_data_dir` - i.e., the properties that are set in the base class' `_set_config_base()` method **and** defined in the schema. In the previous setup they had to be defined *after* this method was called but *before* any path resolution, as they were used in paths. So there was a rigid sequence of method calls: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L224 (see lines 224-230). Such tight coupling, of course, prevents refactoring (and is bad for other reasons).

I think the setting of base config files is a special case and has nothing to do with automatically reading schema and values from kwargs ( here's my thoughts: https://github.com/galaxyproject/galaxy/commit/777379729198a893519ad58f167be761a4ea867d ). And so considering these legit cases when this can happen, I think we can treat it as normal behavior: create attribute unless it exists. Besides, 

I think, the only time that can bite someone, is when you add a new instance variable to the config object and use a name for it that's in the schema (or vise-versa). But that's a remote possibility; besides, it's easily debuggable, especially considering the `create_attr...` method is called so early: https://github.com/galaxyproject/galaxy/blob/163ba59bb189cbb3bc42c7f7404ee075a6da8aeb/lib/galaxy/config/__init__.py#L118 - anything conflicting with this would have to be created *earlier* than that call. In summary, I think treating it as expected behavior is reasonable.

Reply:
Scratch all the above. See latest commit - just check the key against 3 base config attrs and raise exception for anything else.

Reply:
Does it need an `else: log.something()` rather than silently passing now?

Reply:
I don't think so: it's expected behavior and happens on every load:
```
if attribute doesn't exist:
    create attribute - as expected 
else:
    if exists and is NOT a base config:
        raise error
    else: (is a base config)
        do not create attribute - as expected
```

Reply:
Ok, thanks for the clarification